### DISCUSSION TITLE
Suggestion: don't ban immediately

### DISCUSSION THREAD

**Discussion by glebsterx** [umputun/tg-spam#73] [Ideas]
In general, as far as I understand, when a bot detects spam, it bans the user by blacklisting it (a standard Telegram feature)

But sometimes the bot mistakenly bans an ordinary new user, then you have to write a personal message and return the user manually, because often the privacy settings prohibit adding to groups.

_The suggestion is the following - delete messages and limit the rights of suspicious users (rather than blacklisting them) if an admin chat is configured. And let the administrators blacklist them by confirming the ban._
If the chat for administrators is not configured - leave the current behavior - ban from the first message.

**Comment by umputun** [comment]
isn't this what `--training` mode already does?

https://github.com/umputun/tg-spam?tab=readme-ov-file#training-the-bot-on-a-live-system-safely

> In this mode admin can ban users manually by clicking the "confirm ban" button on the message. This allows running the bot as a post-moderation tool and training it on the fly.

**Comment by umputun** [comment]
I'm adding `soft-mode` flag what restircts users instead of the ban. See #74

**Reply by glebsterx** [reply]
in training mode, as I understand it, there is no restriction of the user and deletion of the message, if the soft mode does this - thank you for your responsiveness and efficiency =)

**Reply by umputun** [reply]
yw. This was merged in and released as a part of v1.12.0. Personally, I don't have much use of it, so give it a try and see if it does what you are expecting. The change was trivial, and I have added tests, so it should be working, but who knows ;)

**Reply by glebsterx** [reply]
thanks again - I've already installed it - I'll test it and if I have anything to say I'll be sure to write in this thread

**Comment by glebsterx** [comment]
I partially checked today the new work scheme (at first without a fix - I thought I had configured something wrong, then I saw the v1.12.1 update) - this is what I expected, except for one thing - I expected that after confirmation through the administrators chat user will move from the “exceptions” (users with custom permissions) to the “blacklist” and will be (automatically) removed from the chat.
It has not yet been possible to check false positives and removal from the list of restricted users - spam detection generally works very well =)

**Comment by glebsterx** [comment]
Works as a charm!
![image](https://github.com/umputun/tg-spam/assets/8779304/1c770eb7-8ef1-45e8-b82a-338c796e91d6)
![image](https://github.com/umputun/tg-spam/assets/8779304/5fefc76c-e3e1-4d3a-bf5c-3cd90c0cada5)
![image](https://github.com/umputun/tg-spam/assets/8779304/a8e0a767-adb3-40e3-add3-d965a3a7f44b)

**Reply by proarachnid** [reply]
It's great! Do I understand correctly that if the "soft" + "training" mode is enabled, then there will be no spam in the channel as in the screenshots?

**Reply by glebsterx** [reply]
> "soft" + "training"

To clear a chat of spam, as a rule, an enabled bot is enough, but the soft mode allows you to painlessly return a user whose rights the bot has mistakenly limited.

The last change does not have a version number - it will allow you to manually permanently ban a user if he really sent spam

**Reply by proarachnid** [reply]
I apologize for the incorrect presentation of thoughts.  I didn't mean spam from spammers, but messages listing the restrictions imposed on the user.
**I hate online translators((

**Reply by glebsterx** [reply]
Messages with a list of user rights are from the “Recent Actions” interface of the chat admin in Telegramц

Ну так то можно и по-русски говорить наверное =)
Сообщения со списком прав пользователя это из интерфейса "Недавние действия" админки чата в телеграме

**Reply by proarachnid** [reply]
Большое спасибо за отзывчивость ;)
Теперь все понятно. Просто обратил внимание, что средний скрин как-раз из админского чата, тут вопросов нет, а остальные выглядят так, будто бот в рабочий чат накидал нотисов. По этому и хотел уточнить учитывает ли замечательная функция soft-ban включенный в данный момент training режим что бы не мусорить лишний раз.

**Reply by glebsterx** [reply]
messages in the admin chat do not change and do not depend on the modes - visually nothing changes, but..
in the latest version with soft mode (which does not have a version number) - the "keep it banned" button bans the user completely and adds the message to the list of spam samples (as without a soft ban), and "unban for real" - returns the rights to standard
that is, now we essentially have a purgatory in which everyone who, in the bot’s opinion, has sinned lives =)

This is what I meant in the first message in this thread. Once again, low bow and gratitude to the author on behalf of our chat of almost 3000 people.
______________

сообщения в админском чате не меняются и не зависят от режимов -- визуально ничего не меняется, но..
в последней версии c soft mode (которая не имеет номера версии) -- кнопка "keep it banned" банит пользователя окончательно и добавляет сообщение в список спам-сэмплов (как и без мягкого бана), а "unban for real" -- возвращает права на стандартные
то есть теперь мы имеем по сути чистилище в котором живут все кто по мнению бота согрешил =)

Это то что я имел в виду в первом сообщении в этом треде. Автору еще раз низкий поклон и благодарность от имени нашего чата на почти 3000 человек.

**Reply by umputun** [reply]
tagged as v1.12.2

**Comment by airwavy19** [comment]
## Cutover complete (post-session follow-up)

Per the user's request, re-pointed the live `watchtower.service` unit at the new cs-watchtower repo:

- Stopped: `systemctl --user stop watchtower.service` → `inactive`
- Copied state.json across: `/home/popwavy/Documents/cscustomscanner/backend/watchtower/state.json` → `/home/popwavy/Documents/cs-watchtower/state.json` (mode 600, 10 sites intact)
- Copied .env across (mode 600, same Discord token — `WATCHTOWER_DISCORD_TOKEN` env var not set yet, the bot reads `DISCORD_TOKEN` as fallback and logged in successfully as HachiRelay#8190)
- Generated new unit from `cs-watchtower/infra/systemd/watchtower.service` via sed ({{REPO}} + {{USER}} substitution)
- Installed to `~/.config/systemd/user/watchtower.service` + `systemctl --user daemon-reload`
- `systemctl --user enable --now watchtower.service` → `active (running)`, PID 3892656
- 35s smoke: 1 process (no duplicates), NRestarts=0, scheduler tick running, first probe completed (`5/6 up, 1 down`), alert emitted to channel 1536757566078722118, Discord gateway connected
- Cleaned up the detach worktree

**Left in place (operator may want to delete later):** the orphan `cscustomscanner/backend/watchtower/state.json` (gitignored by c3fbfda). Canonical home is `~/Documents/cs-watchtower/state.json` now.

**Known cosmetic warning in the new unit's log:** `Could not write heartbeat: [Errno 30] Read-only file system: '/tmp/watchtower.heartbeat'` — same as the old unit. `ProtectSystem=strict` in the systemd unit protects /tmp; the heartbeat file is supposed to be `ReadWritePaths=-/tmp/watchtower.heartbeat` but that's the systemd-wide /tmp, not the user's per-process /tmp. The bot handles the missing heartbeat gracefully (warns + continues) — pre-existing behavior, not a regression.

**Comment by airwavy19** [comment]
## Follow-up #2 (2026-08-18 23:00 cscustomscanner + cs-watchtower)

**Two more partitions landed + the new cs-watchtower web app shipped.**

### 1. mmxquota cleanup — PR #194 (merged)

The operator flagged that mmxquota owns the quota work and shouldn't share a repo with cscustomscanner. Audit confirmed:
- `quota-bot.service` (PID 1372613) runs mmxquota's bot, not ours.
- cscustomscanner's `backend/quota_bot/` (1.1M, 18 .py files) + 5 mmxquota-collector-* unit templates + 4 quota tests were dead since the 2026-08-16 inbound migration.
- Stripped all of it; cscustomscanner now consumes quota data through `/api/quota/*` reads of mmxquota's Postgres DB (no module import). The 2026-08-16 migration completed.

Diff: 33 insertions / 13480 deletions across 48 files.

### 2. cs-watchtower web app — pushed to main @ cs-watchtower

Self-contained Flask + vanilla HTML/CSS/JS mobile-first operator dashboard on :9210. Reads from the same `state.json` the bot owns, exposes the same CRUD surface the Discord modal offers, plus two new things:

- **Pi session id announcement in the alert embed.** When a [WORK] site with `path` goes up→down, the bot looks up the most recent Pi session in that path's `~/.pi/agent/sessions-archive/` and resumes it with `WATCHTOWER_PI_PROMPT`. The session id appears in the Discord alert (and the closure embed on recovery) — no out-of-band UI needed.

- **Mobile-first tile UI.** Tiles grouped by folder (`category`), per-tile status badges (Up/Down/Degraded + HTTP code + relative time), bottom-sheet drawer for Add/Edit/Detach/Pin. Runtime fields surface in a sibling `runtime` block so the form can never re-submit a fabricated verdict on a fresh record.

122 tests pass (`python3 -m pytest tests/ -q`). Live unit `watchtower-web.service` installed and running (PID active, 19MB memory, /health returns version+commit+uptime+site counts per AGENTS.md §7).

The bot process also restarted (new PID) to pick up the Pi auto-investigate code. Startup log now shows "Pi auto-investigate: 0/12 work sites have a session mapping" + the list of sites without a `path` field — gap is visible to the operator instead of silently dropped.

### Lessons written

- cs-watchtower repo: Flask + importlib test isolation pattern (state module instances, conftest monkeypatch caveats, tmp_path reuse across runs).
- cscustomscanner: 3-way partition pattern (watchtower / quota_bot / web app) — when one repo gets a microservice merge-in, schedule the merge-out as a follow-up issue immediately, or the inbound code becomes dead within weeks.

### Open follow-ups (still pending)

- Add `pre-push` hook + `check:fast` script to cs-watchtower.
- Decide whether the operator wants the web UI on a public hostname via Cloudflare tunnel (per existing pattern in cscustomscanner/AGENTS.md §18).

**Reply by paskal** [reply]
I think your agent posts nonsense to the wrong project.
